numpyro/hmc_util.py

```diff
@@ -586,10 +586,13 @@ def initialize_model(rng, model, *model_args, **model_kwargs):
         to convert unconstrained HMC samples to constrained values that
         lie within the site's support.
     """
+    dtype = model_kwargs.pop('dtype', np.float32)
     model = seed(model, rng)
     model_trace = trace(model).get_trace(*model_args, **model_kwargs)
     sample_sites = {k: v for k, v in model_trace.items() if v['type'] == 'sample' and not v['is_observed']}
     inv_transforms = {k: biject_to(v['fn'].support) for k, v in sample_sites.items()}
-    init_params = transform_fn(inv_transforms, {k: v['value'] for k, v in sample_sites.items()}, constrain=False)
+    init_params = transform_fn(inv_transforms,
```

Comment:
Are these the only changes needed to run in `float64`?

Reply:
Yeah, we just need init values are float64 in HMC because samplers is not much involved (excet for bernoulli/uniform sampling to make transition/direction). I believe to make SVI run in float64, we have to seed dtype into samplers. So a better solition is to allow dtype keywork in distributions’ sample method. What do you think?

Reply:
Hmm..if distribution parameters are `float64`, the samples drawn should automatically be float64 values, right? I think the most non-invasive way to support `float64` might be to have users explicitly use `np.float64` dtype in their models. We may still need to use the sample's dtype to construct float64 arrays when generating momentum etc, but it will be nice to be as less invasive as possible. Would also be good to convert an example/test to use `float64` and test.

Reply:
> if distribution parameters are float64, the samples drawn should automatically be float64 values

This is the case but e.g. uniform samples take max value is `1 - finfo(float32).eps` instead of `1 - finfo(float64).eps`. Random samplers in JAX has a `dtype` arg where default dtype is np.float32. If distribution params are float64, then operators like `loc + scale * eps` will return `float64`. But if distribution params are python scalar, then the returned value is `float32` (if eps comes from a random sampler with default dtype=np.float32). If we override the default behaviour of JAX samplers by creating wrappers in `util` with default dtype is `np.float64`, then we don't need to make this change in `initialize_model`. I think that this is also a good way to do because it can support both HMC and SVI.

> I think the most non-invasive way to support float64 might be to have users explicitly use np.float64 dtype in their models.

This actually makes things complicated, when users have to rewrite their models to run models in float64. Even if users make dist params to be float64, the samples are still in float32 and then converted to float64. If you don't like to have additional `dtype` arg then I propose to make wrappers as above. Then users/we don't have to touch anything. Just enable float64, the inference will run in float64. Otherwise, it will run in float32. Of course, there is no support for: enable float64 and want to run inference in float32.

> Would also be good to convert an example/test to use float64 and test.

Yes, because we have to enable float64 at beginning of the test, so I guess we need to add separated commands in travis. Something like
```
JAX_ENABLE_x64=True pytest test/test_mcmc.py
```
I will do it. If you have other suggestions, please let me know.

Reply:
> If we override the default behaviour of JAX samplers by creating wrappers in util with default dtype is np.float64, then we don't need to make this change in initialize_model.

I think we should do that anyways (assuming that there isn't another pattern to change the default dtype), since even for HMC, we actually do want the sampler to use `float64` and not just do the computation in `float32` and then cast it to `float64`. I wish there was a simpler way to achieving this though - an analogue of `torch.set_default_tensor_type()` kind of a thing that did not require changing the default of every singe array constructor.

Reply:
Yeah, I wish so too. But jax random samplers do not follow that semantics. Other constructors such as `np.ones()`, `np.arange()`, `np.log(1)` works as expected, except random samplers. >"<